source/common/formatter/substitution_formatter.cc

```diff
@@ -54,5 +54,101 @@ const re2::RE2& SubstitutionFormatParser::commandWithArgsRegex() {
   // clang-format on
 }
 
+JsonFormatBuilder::FormatElements
+JsonFormatBuilder::fromStruct(const ProtobufWkt::Struct& struct_format) {
+  output_.clear();
+
+  formatValueToFormatElements(struct_format.fields());
+  std::string json_piece = std::move(buffer_);
```

Comment:
> nit: early exit here if buffer_.empty().

It will never be empty here. It will at least contains the ending delimiter of map `}`.

The `output_` is a intermediate product of the configuration loading. We will parse the input `Protobuf::Struct` into a list of template string and pre-sanitized json piece. This PR has lots of related description. And this class also have a detailed comment.
See https://github.com/envoyproxy/envoy/pull/35545#issuecomment-2272759820